### ISSUE TITLE
Nested loops and colliding let variables

### ISSUE THREAD

**Issue by lukescott** [6to5/6to5#510]
I hit a case with nested for loops (simplified):
```javascript
for (let i = 0, len = 10; i < len; i++) {
	i;
	for (let i = 0, len = 10; i < len; i++) {
		i;
		for (let i = 0, len = 10; i < len; i++) {
		  i;
		}
	}
}
```
Transpiles to:
```javascript
for (var i = 0, len = 10; i < len; i++) {
  i;
  for (var _i = 0, _len = 10; _i < _len; _i++) {
    _i;
    for (var _i2 = 0, _len2 = 10; _i2 < _len2; _i2++) {
      _i; // should be: _i2;
    }
  }
}
```

I would assume that the inner `i;` should be `_i2;` because the inner `let i` declaration shadows the outer one. This seems to be the case for the second, but not the third.

I was curious what Traceur did and this was the result:
```javascript
  for (var i = 0, len = 10; i < len; i++) {
    i;
    for (var i$__0 = 0, len$__1 = 10; i$__0 < len$__1; i$__0++) {
      i$__0;
      // should be: for (var i$__2 = 0, len$__3 = 10; i$__2 < len$__3; i$__2++)
      for (var i$__2 = 0, len$__3 = 10; i$__0 < len$__1; i$__0++) { 
        i$__0; // should be: i$__2
      }
    }
  }
```

Notice how the inner for loop has the wrong condition and increment, which differs from 6to5. I'm assuming that both of these results are wrong. My assumption could be wrong, but surely both transpilers should be doing the same thing, right?

Here's a different, but possibly similar case:
```javascript
let i = 0;
{
    i;
    let i = i;
    {
      	i;
        let i = i;
    }
}
```
Transpiles to (basically the same for Traceur):
```javascript
var i = 0;
{
  _i; // Should be i;
  var _i = _i; // Should be var _i = i;
  {
    _i2; // Should be _i;
    var _i2 = _i2; // Should be var _i2 = _i;
  }
}
```

Are my assumptions correct? If not, what is the correct output?

**Comment by fkling** [comment]
I actually, if I follow the [spec](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-let-and-const-declarations) correctly, then your *second* example should throw a reference error because you are trying to access a lexically bound variable before it got initialized. Relevant part:

> The variables are created when their containing Lexical Environment is instantiated but may not be accessed in any way until the variable’s LexicalBinding is evaluated. 

References: [Block evaluation](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-block-runtime-semantics-evaluation), [`GetBindingValue`](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-declarative-environment-records-getbindingvalue-n-s).

**Comment by lukescott** [comment]
The second example probably isn't relevant to the first. It was more out of curiosity. Probably should have put that in a separate issue. Although, the second case is valid in other languages.

**Comment by UltCombo** [comment]
@lukescott nope, the identifier counts as uninitialized as you're declaring an identifier with the same name in that scope which control hasn't reached yet -- the identifier is in the temporal dead zone. Likewise:

```
let i = i;
```

Is also a reference error, as `i` is only considered initiliazed after the initializer expression has been evaluated.

**Comment by Alxandr** [comment]
I just ran into what I assume to be the same issue too. There seems to in general be a few problems with `let` scoping. For starters, loops (`for`-loops at the least) do not have the right scoping as far as I know. The following code exemplifies just that:

```js
let a = 5;
let list = [];
{
  let a = 10;
  {
    let a = 20;
    for (let i = 0; i < a; i++) {
      let a = i * 25;
    }
  }
}
```

This code get's transpiled into the following:

```js
"use strict";

var a = 5;
var list = [];
{
  var _a = 10;
  {
    var _a2 = 20;
    for (var i = 0; i < _a3; i++) {
      var _a3 = i * 25;
    }
  }
}
```

This is obviously not what I intended. The loop is suddenly from `i = 0 to _a3`, instead of `_a2`  which it should be. This probably stems from 6to5 ignoring the fact that blocks carries their own scope even when used in loops (and other constructs), so we get something like this:

```
<scope start>for (let i = 0; i < a; i++) <scope start>{
...
}<scope end><scope end>
```

Now, if I add a closure to the inner loop, to force it to generate lambdas to store state, I get some completely crazy results. The following code:

```js
let a = 5;
let list = [];
{
  let a = 10;
  {
    let a = 20;
    for (let i = 0; i < a; i++) {
      let a = i * 25;
      list.push(() => a);
    }
  }
}
```

transpiles into this:

```js
"use strict";

var a = 5;
var list = [];
{
  (function () {
    var _a = 10;
    {
      (function () {
        var _a2 = 20;
        for (var i = 0; i < _a3; i++) {
          (function () {
            var _a3 = i * 25;
            list.push(function () {
              return _a;
            });
          })();
        }
      })();
    }
  })();
}
```

Instead of what I would expect, which is the following:
```js
"use strict";

var a = 5;
var list = [];
{
  var _a = 10;
  {
    var _a2 = 20;
    for (var i = 0; i < _a3; i++) {
      (function() {
        var _a3 = i * 25;
        list.push(function() { return _a3; });
      })();
    }
  }
}
```

I would imagine there are a few good cases to add to the test suite here :)

**Comment by sebmck** [comment]
@Alxandr 

> I would imagine there are a few good cases to add to the test suite here :)

Yep, we currently pass the Traceur block scoping test suite but it's obviously not enough. It probably suffers from the same problem.

Basically everything works swimmingly if you don't have any variable collisions. I've rewritten the let scoping transformer multiple times but the actual fact is that let scoping is **hard**. I appreciate all the bug reports and I've been hesitant to dive into the let scoping transformer code because of how dense it is. Hopefully I'll be able to get some time to fix these collision issues.

**Comment by sebmck** [comment]
If some of you can write up some tests for these issues it'd be extremely appreciated. Basically something like [this](https://github.com/6to5/6to5/blob/master/test/fixtures/transformation/es6-let-scoping/exec-duplicate-in-upper-scope/exec.js) or [this](https://github.com/6to5/6to5/blob/master/test/fixtures/transformation/es6-let-scoping/exec-for-loop-head/exec.js).

**Comment by Alxandr** [comment]
Here's one:

```js
assert.equal(() => {
  let sum = 0;
  let a = 0;
  {
    let a = 10;
    for (let i = 0; i < a; i++) {
      let a = 1;
      sum += a;
    }
  }
  return sum;
}(), 10);
```

Sorry I'm not PRing, I'm basically going to bed, and I'd have to fork and create a PR. I'll get you another with a closure inside it before I turn in though.

**Comment by Alxandr** [comment]
And here's another:

```
assert.equal(() => {
  let sum = 0;
  let a = 0;
  {
    let a = 10;
    for (let i = 0; i < a; i++) {
      let a = 1;
      sum += (() => a)();
    }
  }
  return sum;
}(), 10);
```

I got both of these tests to pass manually simply by changing some of the variable names. The first one I just had to change the one in the loop condition to not be the one initialized inside the loop, though that should probably be easily fixed by just introducing a scope-barrier before the block in the loop. The second I also had to change the value returned by the anonymous method inside the loop.

**Comment by sebmck** [comment]
No PR is fine, easier for me to do it. Thanks!

On Saturday, 17 January 2015, Alxandr <notifications@github.com> wrote:

> And here's another:
>
> assert.equal(() => {
>   let sum = 0;
>   let a = 0;
>   {
>     let a = 10;
>     for (let i = 0; i < a; i++) {
>       let a = 1;
>       sum += (() => a)();
>     }
>   }
>   return sum;
> }(), 10);
>
> I got both of these tests to pass manually simply by changing some of the
> variable names. The first one I just had to change the one in the loop
> condition to not be the one initialized inside the loop, though that should
> probably be easily fixed by just introducing a scope-barrier before the
> block in the loop. The second I also had to change the value returned by
> the anonymous method inside the loop.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/6to5/6to5/issues/510#issuecomment-70359314>.
>


-- 
Sebastian McKenzie

**Comment by sebmck** [comment]
Ok so I've solved most of these issues, I'm confident I'll be able to work out the remaining. I'll keep you posted.

**Comment by UltCombo** [comment]
Nice work @sebmck !
Perhaps you can also import the added tests from google/traceur-compiler#1641

**Comment by lukescott** [comment]
Awesome, thanks @sebmck!

**Comment by sebmck** [comment]
@UltCombo Thanks!
@lukescott No problem!

So I've rewritten a lot of the let scoping transformer. It now passes all the additional Traceur block scoping tests as well as the ones posted in this issue. It should be much more reliable and stable now.

**Comment by sebmck** [comment]
Fixed as of [2.13.0](https://github.com/6to5/6to5/blob/master/CHANGELOG.md#2130).

**Comment by sebmck** [comment]
Please let me know if you run into anymore issues as hopefully they've all been resolved.
